File: Sources/Olaf/UI/Bridge/ExternalToolBridge.swift
import Foundation

/// Represents a transition from the Olaf viewer to another diagnostic tool.
///
/// The package is **not** tied to any external tool: the app conforms its own bridge to this
/// protocol and registers it with `OlafUI.register(_:)`. No button appears in the viewer if no
/// bridge is registered.
public protocol ExternalToolBridge: Sendable {
    /// Title shown in the viewer toolbar.
    var title: String { get }
    /// System symbol name (SF Symbol), optional.
    var systemImage: String? { get }
    /// Opens the tool. Closing the Olaf viewer is the caller's responsibility (usually dismiss first).
    @MainActor func open()
}

public extension ExternalToolBridge {
    var systemImage: String? { nil }
}

/// Process-wide registry of registered external tool bridges. Thread-safe.
final class ExternalToolRegistry: @unchecked Sendable {

    static let shared = ExternalToolRegistry()

    private let lock = NSLock()
    private var bridges: [any ExternalToolBridge] = []

    func register(_ bridge: any ExternalToolBridge) {
        lock.lock(); defer { lock.unlock() }
        bridges.append(bridge)
    }

    func removeAll() {
        lock.lock(); defer { lock.unlock() }
        bridges.removeAll()
    }

    var all: [any ExternalToolBridge] {
        lock.lock(); defer { lock.unlock() }
        return bridges
    }
}

File: Android/olaf/src/main/java/com/olaf/ui/bridge/ExternalToolBridge.kt
package com.olaf.ui.bridge

import android.content.Context

/**
 * Lets the host surface its own diagnostics tool inside the Olaf viewer as a button. Olaf itself
 * is deliberately **not tied to any external tool** — the host writes the bridge and registers it
 * through `OlafUI.register`.
 *
 * ```kotlin
 * OlafUI.register(object : ExternalToolBridge {
 *     override val title = "SomeTool"
 *     override fun open(context: Context) { SomeTool.show(context) }
 * })
 * ```
 */
interface ExternalToolBridge {
    /** Button label shown in the viewer. */
    val title: String

    /** Invoked when the button is tapped. */
    fun open(context: Context)
}

/** Process-wide registry of external tool bridges. */
internal object ExternalToolRegistry {

    private val bridges = mutableListOf<ExternalToolBridge>()

    fun register(bridge: ExternalToolBridge) {
        synchronized(bridges) {
            if (bridges.none { it.title == bridge.title }) bridges.add(bridge)
        }
    }

    fun removeAll() {
        synchronized(bridges) { bridges.clear() }
    }

    fun all(): List<ExternalToolBridge> = synchronized(bridges) { bridges.toList() }
}
